### ISSUE TITLE
Proposal: plugin API

### ISSUE THREAD

**Issue by petehunt** [excalidraw/excalidraw#2189]
There are a ton of features being asked for that may or may not make it into Excalidraw:
* https://twitter.com/Vjeux/status/1310329936796762112
* #2184
* #2170
* #1718
* #1689

and probably many others. These are all good ideas but the core team is small and adding new features requires a lot of discretion so the app doesn't lose its simplicity.

I wonder if we should consider introducing the notion of plugins. I was thinking they could work similar to a DAW workstation's plugins, where we basically give them a rectangle that they can paint with whatever they want for the user interface, and we provide an API for them to modify the `ExcalidrawElement`s. It would require a lot of thought and design up front but I think it would be really neat if we could enable an ecosystem of plugins developed by third parties. We could even take the best plugins and merge them into core if and where it makes sense and after it's been proven to be valuable as a plugin.

If there's any interest in this I'm happy to work on a spec.

**Comment by dai-shi** [comment]
Yeah, I've wanted such API that I can plugin [excalidraw-animate](https://github.com/dai-shi/excalidraw-animate) and [excalidraw-claymate](https://github.com/dai-shi/excalidraw-claymate).

**Comment by excitedbox** [comment]
YES. Also adding features such as text, voice, video chat, asset libraries, an other features may not be needed by everyone so a quick way to add or remove them would be awesome. Using a CDN to include JS libraries you could quickly add or remove features as well in a plugin manager.

**Comment by ReneCode** [comment]
Yes, the ability to add additional features to the core application opens a lot of user-specific use-cases.
The user should define with plugin(s) he/she want's to use.

But what do we want to be plugin-able ? Ideas:
- after each changing of the scene-content (debounced) the whole elements-array is send to the plugin and a modified elements-array will be returned, re-rendering after getting the new elements-array.
- there are some "hooks" in the core-application a plugin can connect to. (hooks for add, remove, change element) only the edited element will be send to the plugin results new/modified elements
- allow adding additional attributes to ExcalidrawElement
- allow adding new Actions (with UI components)

**Comment by ReneCode** [comment]
> Yeah, I've wanted such API that I can plugin [excalidraw-animate](https://github.com/dai-shi/excalidraw-animate) and [excalidraw-claymate](https://github.com/dai-shi/excalidraw-claymate).

What "kind" of API would you need for it ?
May be starting with a real usecase is a good idea.

**Comment by dai-shi** [comment]
@ReneCode Those two have different requirements. Claymate should be easier. It needs to add a new island and have access to some (exported?) functions.

<img width="400" alt="2020-10-27 6 41 07" src="https://user-images.githubusercontent.com/490574/97231856-89ce8d00-181f-11eb-9bd8-794a40a61a40.png">

**Comment by ReneCode** [comment]
![plugin](https://user-images.githubusercontent.com/1595098/97812393-4d68ba00-1c81-11eb-90bf-44b7d8c00a7b.gif)

I've made some tries. 

The plugin is a react-app. It will be rendered in an iframe inside the excalidraw-app.

Data is exchanged via message passing.

plugin -> core: window.parent.postMessage(...)

core -> plugin: iframe.contentWindow.postMessage(...)

Some protocoll has to be defined. I started with: "createSnapshot" that will create 2dContext imageData from the current scene. (as made in [excalidraw-claymate](https://github.com/dai-shi/excalidraw-claymate/tree/master/src))

You can take a look on the plugin here: https://github.com/ReneCode/excalidraw-plugin/blob/master/src/App.js

some questions:
- is iframe message passing the right way?
- where should the plugin been rendered (I choose the right side, optional bottom side ?)
- I started with only one active plugin and choosing is via the URL - good idea ?

**Comment by juandspy** [comment]
any news about this plugin api?

**Comment by Igonato** [comment]
Small excalidraw is usefull when it itself is used as a plugin (i.e. VSCode, Obsidian). Meybe the plugin API should be reserved to a separate project? Like Jupyter/JupyterLab this can be Excalidraw/ExcalidrawStudio with the original project kept simple and narrow-scoped.
